### DISCUSSION TITLE
Renovate widens versions to one intentionally skipped over

### DISCUSSION THREAD

**Discussion by nphilipp** [renovatebot/renovate#20808] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

This is about CentOS/duffy#826.

Right now, the version range for `SQLAlchemy` is configured to be `^1.4.25 || ^2.0.5` for poetry, because all versions from 2.0.0 to 2.0.4 suffer from a regression affecting the project. Renovate is configured to widen version ranges, and it (mistakenly IMO) opened a PR to make this `^1.4.25 || ^2.0.5 || ^2.0.4`.

### What you tried so far.

I've tried reproducing it by adding it as a testcase into `lib/modules/versioning/poetry/index.spec.ts`, but it doesn’t get flagged (but I may be running the tests wrongly):

```
diff --git a/lib/modules/versioning/poetry/index.spec.ts b/lib/modules/versioning/poetry/index.spec.ts
index 7d919a24c..dbf2dad6e 100644
--- a/lib/modules/versioning/poetry/index.spec.ts
+++ b/lib/modules/versioning/poetry/index.spec.ts
@@ -227,6 +227,7 @@ describe('modules/versioning/poetry/index', () => {
     ${'~1'}             | ${'replace'}  | ${'1.2.3'}         | ${'2.0.0'}         | ${'~2'}
     ${'^2.2'}           | ${'widen'}    | ${'2.2.0'}         | ${'3.0.0'}         | ${'^2.2 || ^3.0.0'}
     ${'^2.2 || ^3.0.0'} | ${'widen'}    | ${'3.0.0'}         | ${'4.0.0'}         | ${'^2.2 || ^3.0.0 || ^4.0.0'}
+    ${'^1.4.25 || ^2.0.5'} | ${'widen'} | ${'2.0.5.post1'}   | ${'2.0.5.post1'}   | ${'^1.4.25 || ^2.0.5'}
     ${'^3.5'}           | ${'pin'}      | ${'3.5'}           | ${'3.5'}           | ${'3.5'}
   `(
     'getNewValue("$currentValue", "$rangeStrategy", "$currentVersion", "$newVersion") === "$expected"',
```

### Relevant debug logs

<details><summary>Logs</summary>

```
          {
            "changelogUrl": "https://docs.sqlalchemy.org/en/latest/changelog/",
            "currentValue": "^1.4.25 || ^2.0.5",
            "currentVersion": "2.0.5.post1",
            "datasource": "pypi",
            "depName": "SQLAlchemy",
            "depType": "dependencies",
            "fixedVersion": "2.0.5.post1",
            "homepage": "https://www.sqlalchemy.org",
            "isSingleVersion": false,
            "lockedVersion": "2.0.5.post1",
            "managerData": {
              "nestedVersion": true
            },
            "packageName": "sqlalchemy",
            "registryUrl": "https://pypi.org/pypi",
            "versioning": "poetry",
            "warnings": [],
            "updates": [
              {
                "bucket": "major-2",
                "newVersion": "2.0.4",
                "newValue": "^1.4.25 || ^2.0.5 || ^2.0.4",
                "releaseTimestamp": "2023-02-17T16:55:09.000Z",
                "newMajor": 2,
                "newMinor": 0,
                "updateType": "patch",
                "isRange": true,
                "branchName": "renovate/all-minor-patch"
              }
            ]
          },
```

</details>

**Comment by rarkins** [comment]
This is strange. Does `2.0.4` sort higher than `2.0.5.post1` in our poetry versioning?

We would need either an incorrectly failing unit test or a reproduction to be able to move ahead.

**Reply by nphilipp** [reply]
> This is strange. Does `2.0.4` sort higher than `2.0.5.post1` in our poetry versioning?

I can’t say if it does, only that it shouldn’t 😉. Poetry doesn’t think `2.0.4` is higher than `2.5.0.post1` anyway:

```
(duffy) nils@makake:~/src/fedora-infra/duffy (dev)> grep -1 'name = "sqlalchemy"' poetry.lock 
[[package]]
name = "sqlalchemy"
version = "2.0.5.post1"
(duffy) nils@makake:~/src/fedora-infra/duffy (dev)> poetry update sqlalchemy
Updating dependencies
Resolving dependencies... (1.5s)

No dependencies to install or update
(duffy) nils@makake:~/src/fedora-infra/duffy (dev)>
```

> We would need either an incorrectly failing unit test or a reproduction to be able to move ahead.

I wouldn’t know where to start finding the relevant unit tests in Renovate but I’ll try to cook up a minimal repo exhibiting the issue.

**Reply by nphilipp** [reply]
Alright. I created https://github.com/nphilipp-playground/renovate-poetry-widen-issue-test and added the Renovate app to it. This created the nphilipp-playground/renovate-poetry-widen-issue-test#1 PR which tells me that it would attempt to do the same thing (i.e. append ` || ^2.0.4`). Is this enough for you, or shall I merge the PR so it actually submits the PR with this change?

**Reply by rarkins** [reply]
It should be enough as-is, thanks
